app/src/main/java/org/p2p/wallet/home/ui/crypto/MyCryptoPresenter.kt

```diff
@@ -0,0 +1,195 @@
+import org.p2p.core.utils.orZero
+import org.p2p.core.utils.scaleShort
+import org.p2p.uikit.model.AnyCellItem
+import org.p2p.wallet.common.mvp.BasePresenter
+import org.p2p.wallet.common.ui.widget.actionbuttons.ActionButton
+import org.p2p.wallet.home.analytics.HomeAnalytics
+import org.p2p.wallet.home.model.HomePresenterMapper
+import org.p2p.wallet.home.model.VisibilityState
+import org.p2p.wallet.home.ui.main.HomeInteractor
+import org.p2p.wallet.infrastructure.transactionmanager.TransactionManager
+import org.p2p.wallet.tokenservice.TokenServiceCoordinator
+import org.p2p.wallet.tokenservice.UserTokensState
+import org.p2p.wallet.transaction.model.TransactionState
+
+class MyCryptoPresenter(
+    private val homeInteractor: HomeInteractor,
+    private val homeMapper: HomePresenterMapper,
+    private val connectionManager: ConnectionManager,
+    private val transactionManager: TransactionManager,
+    private val tokenServiceCoordinator: TokenServiceCoordinator,
+    private val analytics: HomeAnalytics,
+) : BasePresenter<MyCryptoContract.View>(), MyCryptoContract.Presenter {
+
+    private var visibilityState: VisibilityState = VisibilityState.Hidden
+    private var cryptoTokensSubscription: Job? = null
+
+    override fun attach(view: MyCryptoContract.View) {
+        super.attach(view)
+        initializeActionButtons()
+        observeCryptoTokens()
+    }
+
+    override fun refreshTokens() {
+        launchInternetAware(connectionManager) {
+            try {
+                tokenServiceCoordinator.refresh()
+            } catch (cancelled: CancellationException) {
+                Timber.i("Loading tokens job cancelled")
+            } catch (error: Throwable) {
+                Timber.e(error, "Error refreshing user tokens")
+                view?.showErrorMessage(error)
+            }
+        }
+    }
+
+    private fun observeCryptoTokens() {
+        cryptoTokensSubscription?.cancel()
+        cryptoTokensSubscription = launch {
+            tokenServiceCoordinator.observeUserTokens()
+                .collect { handleTokenState(it) }
+        }
+    }
+
+    private fun handleTokenState(newState: UserTokensState) {
+        view?.showRefreshing(isRefreshing = newState.isLoading())
+
+        when (newState) {
+            is UserTokensState.Idle -> Unit
+            is UserTokensState.Loading -> Unit
+            is UserTokensState.Refreshing -> Unit
+            is UserTokensState.Error -> {
+                view?.showErrorMessage(newState.cause)
+            }
+            is UserTokensState.Empty -> {
+                view?.showEmptyState(isEmpty = true)
+                handleEmptyAccount()
+            }
+            is UserTokensState.Loaded -> {
+                view?.showEmptyState(isEmpty = false)
+                initializeActionButtons()
+                showTokensAndBalance(
+                    solTokens = newState.solTokens,
+                    ethTokens = newState.ethTokens
```

Comment:
Let's get rid of old functions like `initializeActionButtons()` or `showTokensAndBalance`

Reply:
Can we try to make a single `CryptoState` which can modify the UI?
I want only one channel which edits the UI